WindowsAzureMessaging/WindowsAzureMessaging/SBNotificationHub.m

```diff
@@ -378,19 +378,58 @@ - (void)deleteRegistrationWithName:(NSString *)templateName completion:(void (^)
 - (void)unregisterAllWithDeviceToken:(NSData *)deviceTokenData completion:(void (^)(NSError *))completion {
     if (deviceTokenData == nil) {
         if (completion) {
-            completion([SBNotificationHubHelper errorForNullDeviceToken]);
+            dispatch_async(dispatch_get_main_queue(), ^{
+                completion([SBNotificationHubHelper errorForNullDeviceToken]);
+            });
         }
 
         return;
     }
 
     NSString *deviceToken = [self convertDeviceToken:deviceTokenData];
-
-    SBThreadParameter *parameter = [[SBThreadParameter alloc] init];
-    parameter.deviceToken = deviceToken;
-    parameter.completion = completion;
-    parameter.isMainThread = [[NSThread currentThread] isMainThread];
-    [self performSelectorInBackground:@selector(deleteAllRegistrationThread:) withObject:parameter];
+    
+    [self retrieveAllRegistrationsWithDeviceToken:deviceToken completion:^(NSArray *registrations, NSError *error) {
+        
+        if (error) {
+            dispatch_async(dispatch_get_main_queue(), ^{
+                if (completion) {
+                    completion(error);
+                }
+            });
+            return;
+        }
+        
+        if (registrations.count > 0) {
+            dispatch_group_t group = dispatch_group_create();
+            __block NSError *innerError = nil;
+            
+            for (SBRegistration *reg in registrations) {
+                NSString *name = [SBNotificationHubHelper nameOfRegistration:reg];
+                dispatch_group_enter(group);
+                [self deleteRegistrationWithName:name completion:^(NSError *err) {
+                    innerError = err;
```

Comment:
Curious, we are deleting several registrations, the completion cal will set the innerError, since we are using completion blocks in a group, those are executed in order BUT what happens if one of the removals does have an error and is followed by a second block:

1. Block A executes, sets innerError
2. Bock B executes, sets innerError, steps on previous value.

Are we interested in keeping track of all the errors or just the last one?

Reply:
We should probably break after the first and not attempt the second